### PR TITLE
Fix handling of `Unknown` nodes in constant optimization

### PR DESCRIPTION
**Pull request by larsk21**

### CODE REVIEW

**Comment by g4rgoil** [comment]
Will this optimize something like `Add(anyNode, Unknown)`?

**Comment by larsk21** [comment]
> Will this optimize something like `Add(anyNode, Unknown)`?

Yes, this is demonstrated in the added uni test. `y + 5` with an unknown `y` is changed to an `Unknown` node.

File: src/main/java/edu/kit/compiler/optimizations/ConstantOptimization.java [hunk 1]
```diff
@@ -126,28 +133,50 @@ private boolean transformMemoryProj(Proj node) {
...
+
+            if (predValue.isUnknown()) {
+                predValue = TargetValueLatticeElement.constant(undefinedCondStrategy.chooseCondValue((Cond)pred));
+                nodeValues.put(pred, predValue);
+            }
+
+            if (predValue.isConstant()) {
+                if (
+                    (node.getNum() == pn_Cond.pn_Cond_false.val && predValue.getValue().equals(TargetValue.getBFalse())) ||
+                    (node.getNum() == pn_Cond.pn_Cond_true.val && predValue.getValue().equals(TargetValue.getBTrue()))
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/ConstantOptimization.java:147]: Is there a reason you're not using `Cond.pn{True,False}`?

**Review by g4rgoil — APPROVED** [review]
Seems to fully resolve the issue of `Unknown` nodes. Maybe we should note somewhere that our instruction selection is not guaranteed to work for graphs where this optimization has not been applied.

File: src/main/java/edu/kit/compiler/optimizations/ConstantOptimization.java (hunk 1 above)
Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/ConstantOptimization.java:147]: No, I just copied it from where I used this before :D

**Comment by larsk21** [comment]
> Seems to fully resolve the issue of `Unknown` nodes. Maybe we should note somewhere that our instruction selection is not guaranteed to work for graphs where this optimization has not been applied.

I added some documentation and a safety-check (which does not resolve the problem though). Maybe we could add a pattern for `Cond` nodes with an `Unknown` node as the selector, just to be safe.

**Comment by g4rgoil** [comment]
> I added some documentation and a safety-check (which does not resolve the problem though). Maybe we could add a pattern for `Cond` nodes with an `Unknown` node as the selector, just to be safe.

How should a `Cond` with `Unknown` selector be handled during instruction selection in your opinion? Just generate an unconditional jump?

**Comment by larsk21** [comment]
> > I added some documentation and a safety-check (which does not resolve the problem though). Maybe we could add a pattern for `Cond` nodes with an `Unknown` node as the selector, just to be safe.
> 
> How should a `Cond` with `Unknown` selector be handled during instruction selection in your opinion? Just generate an unconditional jump?

For example, anything that generates valid code would be acceptable. This way the instruction selection would be independent of the optimizations.

**Comment by g4rgoil** [comment]
> > > I added some documentation and a safety-check (which does not resolve the problem though). Maybe we could add a pattern for `Cond` nodes with an `Unknown` node as the selector, just to be safe.
> > 
> > 
> > How should a `Cond` with `Unknown` selector be handled during instruction selection in your opinion? Just generate an unconditional jump?
> 
> For example, anything that generates valid code would be acceptable. This way the instruction selection would be independent of the optimizations.

Ok, I can implement a pattern for that.

**Pull request merged**